### ISSUE TITLE
(mini.files) preview is unusable with large minified files

### ISSUE THREAD

**Issue by fbuchlak** [echasnovski/mini.nvim#489]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)
- [X] I have updated 'mini.nvim' to latest version

### Module(s)

mini.files

### Description

Plugin explicitly starts treesitter causing neovim to freeze

### Neovim version

0.9.1

### Steps to reproduce

1. install treesitter syntax for js (doesn't matter if it's js/css or any other language)
2. enable mini.files preview
3. create following structure (use at least ~500kB minified js file) i.e.:
```bash
mkdir some-dir
touch some-dir/a.js some-dir/b.js some-dir/d.js
curl https://cdn.jsdelivr.net/npm/angular2@2.0.0-beta.21/bundles/angular2-all.umd.min.js > some-dir/c-slow.js
```
4. browse this structure from `some-dir/a.js` to `some-dir/d.js` with `j`

### Expected behavior

It should be possible to preview a file without starting treesitter / enabling syntax highlighting or decide whenever to show preview for specific file. 

There are few options for current implementation

1. `boolean` option to disable syntax highlighting globally for mini.files
2. `function` option to disable preview completely per file (similarly how non text files work)
3. `function` option to disable syntax highlighting per file

### Actual behavior

Neovim freezes when large minified file is previewed.

It is possible to use `MiniFilesBufferUpdate` to disable syntax and stop treesitter, but you will still experience little freeze first time you preview file (caused by already started treesitter).

```lua
-- current workaround
vim.api.nvim_create_autocmd("User", {
    group = vim.api.nvim_create_autocmd("mygroup", { clear = true }),
    pattern = "MiniFilesBufferUpdate",
    callback = function(args)
        local buffer = args.data.buf_id

        vim.bo[buffer].syntax = nil
        vim.treesitter.stop(buffer)
    end,
})
```

**Comment by echasnovski** [comment]
Thanks for creating this!

I can partially reproduce: it doesn't exactly "freezes" (which I perceived as "indefinitely stopped responding"), but is being blocked for rendering lines.

I'll think about the best way to counter this.

**Comment by fbuchlak** [comment]
> I can partially reproduce: it doesn't exactly "freezes" (which I perceived as "indefinitely stopped responding"), but is being blocked for rendering lines.

It's same for me. Sorry for inaccurate word.

Another workaround could be shortening lines to length of `width_preview`, however this can easily break highlighted syntax.

```lua
-- Compute lines. Limit number of read lines to work better on large files.
local has_lines, read_res = pcall(vim.fn.readfile, path, '', vim.o.lines)
-- - Make sure that lines don't contain '\n' (might happen in binary files).
local lines = has_lines and vim.split(table.concat(read_res, '\n'), '\n') or {}
for i, line in pairs(lines) do
  lines[i] = line:sub(1, opts.windows.width_preview)
end
```

**Comment by echasnovski** [comment]
> > I can partially reproduce: it doesn't exactly "freezes" (which I perceived as "indefinitely stopped responding"), but is being blocked for rendering lines.
> 
> It's same for me. Sorry for inaccurate word.
> 
> Another workaround could be shortening lines to length of `width_preview`, however this can easily break highlighted syntax.
> 
> ```lua
> -- Compute lines. Limit number of read lines to work better on large files.
> local has_lines, read_res = pcall(vim.fn.readfile, path, '', vim.o.lines)
> -- - Make sure that lines don't contain '\n' (might happen in binary files).
> local lines = has_lines and vim.split(table.concat(read_res, '\n'), '\n') or {}
> for i, line in pairs(lines) do
>   lines[i] = line:sub(1, opts.windows.width_preview)
> end
> ```

Yes, it will break highlighting. I'd probably not highlight if total size of lines is above certain threshold.

**Comment by fbuchlak** [comment]
> I'd probably not highlight if total size of lines is above certain threshold.

This is how I've fixed it for now with `MiniFilesBufferUpdate`.
```lua
callback = function(args)
    local bufnr = args.data.buf_id
                                                                      
    if vim.api.nvim_buf_is_loaded(bufnr) then
        local lines = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
        for _, line in pairs(lines) do
            if #line > 3000 then
                vim.treesitter.stop(bufnr)
                vim.bo[bufnr].syntax = nil
                                                                      
                break
            end
        end
    end
end,
```

For me it would be optimal to run this check before starting treesitter / syntax highlighting.

Something like this would be enough
```lua
-- Add highlighting on Neovim>=0.8 which has stabilized API
if vim.fn.has('nvim-0.8') == 1 and opts.windows.preview_syntax_start(buf_id, lines) then -- or just buf_id
```

**Comment by echasnovski** [comment]
What I have in mind is to check that `total_bytes = vim.api.nvim_buf_call(buf_id, function() return vim.fn.line2byte('$') end)` doesn't exceed certain value. Say 1000000 (1 Mb) or 100000 (100 Kb).

**Comment by fbuchlak** [comment]
> What I have in mind is to check that `total_bytes = vim.api.nvim_buf_call(buf_id, function() return vim.fn.line2byte(vim.fn.line('$') + 1) end)` doesn't exceed certain value. Say 1000000 (1 Mb) or 100000 (100 Kb).

Seems better, however shouldn't it be up to user to decide if he wants to enable syntax highlighting? You might want to disable it for specific languages etc. It doesn't have to do anything with treesitter being slow.
Another thing to consider is hiding contents based on user defined conditions (could be .env files and so on). It would be nice to have a possibility to override `buffer_update_file`.

**Comment by echasnovski** [comment]
> > What I have in mind is to check that `total_bytes = vim.api.nvim_buf_call(buf_id, function() return vim.fn.line2byte(vim.fn.line('$') + 1) end)` doesn't exceed certain value. Say 1000000 (1 Mb) or 100000 (100 Kb).
> 
> Seems better, however shouldn't it be up to user to decide if he wants to enable syntax highlighting? You might want to disable it for specific languages etc. It doesn't have to do anything with treesitter being slow. Another thing to consider is hiding contents based on user defined conditions (could be .env files and so on). It would be nice to have a possibility to override `buffer_update_file`.

It feels like too much customization for my taste. I currently feel that having proper syntax highlighting (if it doesn't introduce visible overhead) is always better than having none.

If utilizing `MiniFilesBufferUpdate` works, then it should be fine.

**Comment by echasnovski** [comment]
This should be resolved on latest `main` via limits on when highlighting is started. See [this part of help](https://github.com/echasnovski/mini.nvim/blob/b94d010c61ba46474af78536873db89492b251a4/doc/mini-files.txt#L614-L616).
